src/librustc_mir/transform/const_prop.rs

```diff
@@ -434,32 +434,24 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
     ) -> Option<Const<'tcx>> {
         let span = source_info.span;
 
-        // if this isn't a supported operation, then return None
-        match rvalue {
-            Rvalue::Repeat(..) |
-            Rvalue::Aggregate(..) |
-            Rvalue::NullaryOp(NullOp::Box, _) |
-            Rvalue::Discriminant(..) => return None,
-
-            Rvalue::Use(_) |
-            Rvalue::Len(_) |
-            Rvalue::Cast(..) |
-            Rvalue::NullaryOp(..) |
-            Rvalue::CheckedBinaryOp(..) |
-            Rvalue::Ref(..) |
-            Rvalue::UnaryOp(..) |
-            Rvalue::BinaryOp(..) => { }
-        }
+        let overflow_check = self.tcx.sess.overflow_checks();
 
-        // perform any special checking for specific Rvalue types
-        if let Rvalue::UnaryOp(op, arg) = rvalue {
-            trace!("checking UnaryOp(op = {:?}, arg = {:?})", op, arg);
-            let overflow_check = self.tcx.sess.overflow_checks();
+        // Perform any special checking for specific Rvalue types.
+        // Generally, checks here fall into one of two categories:
+        //   1. Doing additional validation to provide useful lints to the user
+        //        - In this case, we will do some validation and then fall through to the
+        //          end of the function which evals the assignment.
+        //   2. Working around ICE's in miri because we eagerly evaluate MIR that could
+        //      actually be UB if it were to actually run at runtime.
+        //        - In this case, we'll return `None` from this function to stop evaluation.
```

Comment:
Is there still any of that? You fixed once case of ICE.

Reply:
There's the dead-local-access ICE, but I thought that's not an ICE any more with the new machine hook.

Reply:
Yes (https://github.com/rust-lang/rust/pull/64890#discussion_r334712539)

Reply:
Might be worth pointing out that the local check is the only branch in the 2nd category. Is that even correct?

Reply:
Ok, so I left the main comment because I think the overall exposition is helpful but for each branch I added a comment that says whether it falls into category A or B.

Reply:
That's great, thanks. :)